### PR TITLE
[Airflow-1325] Add Elasticsearch log handler and reader

### PR DESCRIPTION
**Pull request by KevinYang21**
### JIRA
- [ ] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW-1325) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1325

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:
- Add ElasticSearch log handler which writes to local file and reads from ElasticSearch (need seperate EFK/ELK stack to have the local file uploaded to ElasticSearch).
- Add support of auto-tailing task instance log, when user scroll to the bottom.

A simple high level overview of the architecture change of the task instance log loading:
old_arch
new_arch

In the new architecture, metadata would be used to mark end of log, current offset in stream, etc.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
tests/utils/log/test_es_task_handler.py
tests/www/test_views.py: test_get_logs_with_metadata

Similar logic has been running in our branch for more than half a year. Tested Elasticsearch as logging backend on our cluster end to end.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by KevinYang21** [comment]
Hi @bolkedebruin and @Fokko, I'm drafting this change to add Elasticsearch log handler. Before iterating on it too much, I think it would be better to run the high level idea of this change by you guys. Please let me know if there's major concern on this approach.

Major changes:

1. No log will be loaded before rendering ti_log.html, instead the html will make calls, potentially repeatedly, to new get_logs_with_metadata() endpoint to load logs (see the graphs in description).
2. Add auto-tailing in ti_log.html.
3. FileTaskHandler().read() would now take metadata as an option parameter to enable streaming read and return (logs, metadatas) tuple, which may contain end of log and next offset info. 
4. New ElasticsearchTaskHandler which writes to local file and reads from Elasticsearch in a streaming way.

Your feedback would be very valuable and highly appreciated!

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -447,7 +447,6 @@ def run(args, dag=None):
         with redirect_stdout(ti.log, logging.INFO),\
                 redirect_stderr(ti.log, logging.WARN):
             _run(args, dag, ti)
-        logging.shutdown()
```
Comment (KevinYang21) [inline airflow/bin/cli.py]: shutdown() is registered in atexit()

File: airflow/models.py [hunk 2]
```diff
@@ -862,7 +862,7 @@ def __init__(self, task, execution_date, state=None):
         self.init_on_load()
         # Is this TaskInstance being currently running within `airflow run --raw`.
         # Not persisted to the database so only valid for the current process
-        self.is_raw = False
+        self.raw = False
```
Comment (KevinYang21) [inline airflow/models.py:860]: only self.raw is used and set outside constructor.

File: airflow/models.py [hunk 3]
```diff
@@ -1945,8 +1945,8 @@ def init_run_context(self, raw=False):
         """
         Sets the log context.
         """
-        self._set_context(self)
         self.raw = raw
+        self._set_context(self)
```
Comment (KevinYang21) [inline airflow/models.py:1960]: raw is used by custom log handler when setting context, e.g. S3TaskHandler use it to determine if it should upload log on close.

File: airflow/utils/log/es_task_handler.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            return
+
+        if not self.mark_end_on_close:
+            return
+
+        # Mark the end of file using EOT character,
+        # so we know where to stop while auto-tailing.
+        if self.handler is not None and self.handler.stream is None:
+            self.handler.stream = self.handler._open()
+            self.handler.stream.write(u'\u0004')
```
Comment (KevinYang21) [inline airflow/utils/log/es_task_handler.py]: Using END OF TRANSMISSION character for now, suggestion are highly appreciated.

File: airflow/www/views.py [hunk 5]
```diff
@@ -719,6 +718,66 @@ def rendered(self):
             form=form,
             title=title, )
 
+    @expose('/get_logs_with_metadata')
+    @login_required
+    @wwwutils.action_logging
+    @provide_session
+    def get_logs_with_metadata(self, session=None):
```
Comment (KevinYang21) [inline airflow/www/views.py:725]: Not putting the endpoint into API to reuse the dagbag. Open for discussion on this.

**Comment by Fokko** [comment]
I'll need some time to look into this. I like the idea.

**Comment by KevinYang21** [comment]
@aoen @saguziel PTAL

**Comment by KevinYang21** [comment]
Note that codes under /tests/utils/log/elasticmock are mostly copy/paste from [ElasticMock](https://github.com/vrcmarcos/elasticmock) with some compatibility changes and License headers added.

**Comment by Fokko** [comment]
This PR should be to apache/master, right?

**Comment by KevinYang21** [comment]
@Fokko it should, was trying to avoid using CI resources from apache while iterating on the PR, thought I would be able to change the base fork in the PR but turns out I cannot. I'll create another PR against apache/master and tag you. Thank you!

**Comment by Fokko** [comment]
Cool, makes sense. Cheers

**Pull request closed without merging**